Added ssh tutorial (#44)
netgroup-polito/crownlabs-site: content/resources/crownlabs_ssh/crownlabs_ssh.md

   ```
   ssh-keygen
   ```

   You can use the `-b` option to specify the length (bit size) of the key, as shown in the following example:

   ```
   ssh-keygen -b 2048
   ```

2. The command will prompt you to enter the path to the file where you want to save the key. A default path and file name similar to `/home/user_name/.ssh/id_rsa` (`C:\Users\username\.ssh\id_rsa` on **windows**) are suggested in parentheses. To accept the default path and file name (_suggested_), press **Enter**. Otherwise, enter the required path and file name, and then press **Enter**. Beware that from now on, this tutorial will assume you accepted the default settings.

3. A password will be asked. You are free to insert one or to ignore it.

Now you will find in the previously chosen path (`$HOME/.ssh` or `C:\Users\%USERPROFILE%\.ssh` if you kept the default one) a file with a **.pub** extension (`id_rsa.pub` as default). This is the file containing your **SSH public key**.\
Note that the directory contains a file called `id_rsa` (without **.pub**), it is the **SSH private key** which you must not share with others.\

> Print the content of your **SSH public key** running `cat $HOME/.ssh/id_rsa.pub` or opening `C:\Users\%USERPROFILE%\.ssh` with an editor (i.e., **notepad**) in **windows**

Reviewer: ```suggestion
Print the content of your **SSH public key** running `cat $HOME/.ssh/id_rsa.pub` or opening `%USERPROFILE%\.ssh` with an editor (i.e., **notepad**) in **windows**
```
USERPROFILE is the equivalent of HOME, it already includes the complete path

<br>

## Connect to a VM from terminal

<br>

1. Create an instance of a VM in your [**CrownLabs Dashboard page**](https://preprod.crownlabs.polito.it/app).

![image-20211112191418307](./img/2_1.png)

<br>

2. Now move to the **[Active page](https://preprod.crownlabs.polito.it/app/active)** and click on the **SSH** icon inside the row of the **VM** that you have just created. A popup with a shell command inside will be shown, **copy** that command.

![image-20211112193803509](./img/2_2.png)

<br>

3. Now open a **terminal**, paste the previously copied command and run it.
   A prompt asking if you trust the host you are trying to connect with will appear. Type **yes** and press **Enter**.
   Then another prompt asking for a **password** might appear. Type **crownlabs** and press **Enter** to **login** inside your **VM**.

![image-20211112194835635](./img/2_3.png)

<br>

## Conclusion

Now you are able to log in your **VMs** using **SSH**. In case you want to access from different PCs you can add other keys following the same process.

If you are interested in an advanced tutorial about **SSH key** you can check this [link](https://docs.gitlab.com/ee/ssh/). 

## How it works

> What allows you to connect is the **bastion**, as shown in the figure below. First of all **SSH client** will try to connect to the bastion (using the option **-J** which specify a **ProxyJump**) and to authenticate using the previously generated **private key**. The **bastion** will check the **client**'s authentication attempt using the **public key** previously shared using the **CrownLabs Dashboard**.

Reviewer 2: ```suggestion
All CrownLabs VMs are characterized by **private IP addresses**, and cannot be accessed directly through SSH.
As shown in the figure below, it is necessary to first establish the SSH connection through a **bastion**, which instead is publicly accessible, and then **jump** to the final destination (i.e., the VM).

Through the `ssh -J bastion@crownlabs.polito.it crownlabs@10.1.2.3` command, the local **SSH client** first connects to the bastion (indeed, the **-J** option specifies a **ProxyJump**), leveraging the generated **SSH key-pair** for authentication purposes.
Specifically, the **bastion** checks the **client**'s authentication attempt using the **public key** previously uploaded to the **CrownLabs Dashboard**.
```

> Now if the **bastion** has authenticated you correctly you will be redirect to your **VM**. Infact the syntax of the end of the command is the classical one **user@ipadress** where the user will be **crownlabs** (the default user on each **VM**) and the **ip** is your **VM's ip**. The **VM** will authenticate you again using your **public and private keys** as the bastion have done before. The **VM** might ask a password in addition (which is **crownlabs**)

Reviewer 2: ```suggestion
If authentication succeeds, you are then redirected to your **VM**. 
Indeed, the second part of the command is the standard **user@ipadress**, where **user** is **crownlabs** (the default user on each **VM**) and **ip** is your **VM's ip**. 
The **VM** will authenticate you again using your **SSH key-pair**, as the bastion previously did.
Alternatively, it might fallback to password authentication (which is **crownlabs**).
```

+++
fragment = "content"
#disabled = true
date = "2021-11-12"
weight = 100
# title = "Connecting to CrownLabs with SSH"
+++

## Introduction

**CrownLabs** allows users to access their VMs through a local SSH client (i.e., executed on their local machines).

This guide shows how to generate a SSH key pair on your operating systems (Windows, Linux, MacOS), how to load it on **CrownLabs** and access your **CrownLabs**' VM using SSH from a local terminal.

Note that the **generation** of the **SSH key pair** and the upload of the **SSH public key** does not have to be repeated each time you create a new **VM**.
This process only needs to be done once per **device.**

## How it works

All CrownLabs VMs are characterized by **private IP addresses**, and cannot be accessed directly through SSH.
As shown in the figure below, it is necessary to first establish the SSH connection through a **bastion**, which instead is publicly accessible, and then **jump** to the final destination (i.e., the VM).

> Through the `ssh -J bastion@crownlabs.polito.it crownlabs@10.1.2.3` command, the local **SSH client** first connects to the bastion (indeed, the **-J** option specifies a **ProxyJump**), leveraging the generated **SSH key-pair** for authentication purposes.

Reviewer 3: ```suggestion
This can be achieved with the `ssh -J bastion@crownlabs.polito.it crownlabs@10.1.2.3` command, in which the local **SSH client** first connects to the bastion (indeed, the **-J** option specifies a **ProxyJump**), leveraging the generated **SSH key-pair** for authentication purposes.
```

Specifically, the **bastion** checks the **client**'s authentication attempt using the **public key** previously uploaded to the **CrownLabs Dashboard**.

If authentication succeeds, you are then redirected to your **VM**. 

> Indeed, the second part of the command is the standard **user@ipadress**, where **user** is **crownlabs** (the default user on each **VM**) and **ip** is your **VM's ip**. 

Reviewer 3: ```suggestion
Indeed, the second part of the command is the standard **user@ipadress**, where **user** is **crownlabs** (the default user on each **VM**) and **ip** is your **VM's IP address**. 
```

> The **VM** will authenticate you again using your **SSH key-pair**, as the bastion previously did.

Reviewer 3: ```suggestion
The **VM** will authenticate you through your **SSH key-pair**, as the bastion previously did.
```

> Alternatively, it might fallback to password authentication (which is **crownlabs**).

Reviewer 3: ```suggestion
Alternatively, it might fallback to traditional password authentication (which is **crownlabs**).
```

![](./img/3_1.png)

## How to generate an SSH key pair

1. Open a terminal and run the following command to create a [**RSA**](<https://en.wikipedia.org/wiki/RSA_(cryptosystem)>) key pair:

   ```
   ssh-keygen
   ```

>    You can use the `-b` option to specify the length (bit size) of the key, as shown in the following example:

Reviewer 3: ```suggestion
   You can use the `-b` option to specify the length (in bits) of the key, as shown in the following example:
```

   ```
   ssh-keygen -b 2048
   ```

2. The command will prompt you to enter the path to the file where you want to save the key. A default path and file name similar to `/home/user_name/.ssh/id_rsa` (`C:\Users\username\.ssh\id_rsa` on **windows**) are suggested in parentheses. To accept the default path and file name (_suggested_), press **Enter**. Otherwise, enter the required path and file name, and then press **Enter**. Beware that from now on, this tutorial will assume you accepted the default settings.

3. A password will be asked. You are free to insert one or to ignore it.

> Now you will find in the previously chosen path (`$HOME/.ssh` or `C:\Users\%USERPROFILE%\.ssh` if you kept the default one) a file with a **.pub** extension (`id_rsa.pub` as default). This is the file containing your **SSH public key**.\

Reviewer 3: ```suggestion
Now you will find in the previously chosen path (`$HOME/.ssh` or `C:\Users\%USERPROFILE%\.ssh` if you kept the default one) a file with a **.pub** extension (`id_rsa.pub` as default). This file contains your **SSH public key**.
```

> Note that the directory contains a file called `id_rsa` (without **.pub**), it is the **SSH private key** which you must not share with others.\

Reviewer 3: ```suggestion
Note that the directory contains a file called `id_rsa` (without **.pub**), it is the **SSH private key** which must not be shared with others.\
```

> Print the content of your **SSH public key** running `cat $HOME/.ssh/id_rsa.pub` or opening `%USERPROFILE%\.ssh\id_rsa.pub` with an editor (i.e., **notepad**) in **windows**

Reviewer 3: ```suggestion
Print the content of your **SSH public key** typing `cat $HOME/.ssh/id_rsa.pub` (Linux) or opening `%USERPROFILE%\.ssh\id_rsa.pub` with an editor such as **notepad** (**Windows**)
```

## Upload your key on CrownLabs

Now open your page on **CrownLabs** and go to the [**Account page**](https://preprod.crownlabs.polito.it/app/account).

<br>

1. In the central box you will see a tab called **SSH key**. Click it to navigate to the **SSH keys** **management** panel.

![image-20211112184900823](./img/1_1.png)

<br>

2. Now click on **Add SSH key** to open a popup that will allow you to upload your key.

![image-20211112185058163](./img/1_2.png)

<br>

3. Paste your **Public key** (previously copied) inside the popup and add a **Name** to later identify it (whatever you want).

![image-20211112190017165](./img/1_3.png)

<br>

4. Now click on **Save**, the popup will close and your **SSH public key** will be uploaded.

![image-20211112190234270](./img/1_4.png)

> <br>

Reviewer 3: ```suggestion
When completing this process, CrownLabs will **push automatically** the above SSH key in all the _console-based_ VMs you created, as well in the Bastion host; therefore you will be able to connect to all the above VMs using your key.
<br>
```

## Connect to a VM from terminal

<br>

1. Create an instance of a VM in your [**CrownLabs Dashboard page**](https://preprod.crownlabs.polito.it/app).

![image-20211112191418307](./img/2_1.png)

<br>

2. Now move to the **[Active page](https://preprod.crownlabs.polito.it/app/active)** and click on the **SSH** icon inside the row of the **VM** that you have just created. A popup with a shell command inside will be shown, **copy** that command.

![image-20211112193803509](./img/2_2.png)

<br>

3. Now open a **terminal**, paste the previously copied command and run it.
   A prompt asking if you trust the host you are trying to connect with will appear. Type **yes** and press **Enter**.
   Then another prompt asking for a **password** might appear. Type **crownlabs** and press **Enter** to **login** inside your **VM**.

![image-20211112194835635](./img/2_3.png)

<br>

## Conclusion

Now you are able to log in your **VMs** using **SSH**. In case you want to access from different PCs you can add other keys following the same process.

> If you are interested in an advanced tutorial about **SSH key** you can check this [link](https://docs.gitlab.com/ee/ssh/). 

Reviewer 3: ```suggestion
If you are interested in an advanced tutorial about **SSH key** you can read this [page](https://docs.gitlab.com/ee/ssh/) on Gitlab.
```

+++
fragment = "content"
#disabled = true
date = "2021-11-12"
weight = 100
# title = "Connecting to CrownLabs with SSH"
+++

## Introduction

**CrownLabs** allows users to access their VMs through a local SSH client (i.e., executed on their local machines).

> In this guide you are going to learn how to generate a SSH key pair on your operating systems (Windows, Linux, MacOS), how to load it on **CrownLabs** and access your **CrownLabs**' VM using SSH from a local terminal.

Reviewer 3: ```suggestion
This guide shows how to generate a SSH key pair on your operating systems (Windows, Linux, MacOS), how to load it on **CrownLabs** and access your **CrownLabs**' VM using SSH from a local terminal.
```

Note: keep the text as short as possible.

Author: I changed it.

   ```
   ssh-keygen -b 2048
   ```

2. The command will prompt you to enter the path to the file where you want to save the key. A default path and file name similar to `/home/user_name/.ssh/id_rsa` (`C:\Users\username\.ssh\id_rsa` on **windows**) are suggested in parentheses. To accept the default path and file name (_suggested_), press **Enter**. Otherwise, enter the required path and file name, and then press **Enter**. Beware that from now on, this tutorial will assume you accepted the default settings.

3. A password will be asked. You are free to insert one or to ignore it.

Now you will find in the previously chosen path (`$HOME/.ssh` or `C:\Users\%USERPROFILE%\.ssh` if you kept the default one) a file with a **.pub** extension (`id_rsa.pub` as default). This is the file containing your **SSH public key**.\
Note that the directory contains a file called `id_rsa` (without **.pub**), it is the **SSH private key** which you must not share with others.\
Print the content of your **SSH public key** running `cat $HOME/.ssh/id_rsa.pub` or opening `C:\Users\%USERPROFILE%\.ssh` with an editor in **windows**
## Upload your key on CrownLabs

> 

Reviewer 4: I suggest:

## Upload your key on CrownLabs

To use SSH with Crownlabs, copy the contents of your public key file to your Crownlabs account.

You can do this manually or use a script. For example, to copy an RSA key to the clipboard:

 **macOS**:

```bash
 tr -d '\n' < ~/.ssh/id_rsa.pub | pbcopy
```

    **Linux (requires the xclip package)**:

```bash 
xclip -sel clip < ~/.ssh/id_rsa.pub
```

   **Git Bash on Windows**:

```bash
 cat ~/.ssh/id_rsa.pub | clip
```

    Replace `id_rsa.pub` with your filename. For example, use `id_ed25519.pub` for ED25519.

In contrast, you can print the content of your **SSH public key** running `cat $HOME/.ssh/id_rsa.pub` or opening `C:\Users\%USERPROFILE%\.ssh` with an editor in
**windows** and then copy it as well.

Reviewer 2: Personally agree with @cheina97. What @Ralls0 proposed is really cool, but probably overkill for the target audience (if a command for any reason fail -- e.g., a package is missing -- then they would not know what to do next). Would keep this tutorial as simple as possible.

Nonetheless, would discuss this during the meeting today.

+++
fragment = "content"
#disabled = true
date = "2021-11-12"
weight = 100
# title = "Connecting to CrownLabs with SSH"
+++

## Introduction

> [**CrownLabs**](https://crownlabs.polito.it/) allows users access via SSH to their VMs using SSH directly on their machines.

Reviewer 2: ```suggestion
[**CrownLabs**](https://crownlabs.polito.it/) allows users to access their VMs through a local SSH client (i.e., executed on their local machines).
```

> In this guide you are going to learn how to generate a key on the main operative systems (Windows, Linux, MacOS), how to load it on [**CrownLabs**](https://crownlabs.polito.it/) and log in your [**CrownLabs**](https://crownlabs.polito.it/)'VM using SSH from terminal.

Reviewer 2: ```suggestion
In this guide you are going to learn how to generate a SSH key pair on your operating systems (Windows, Linux, MacOS), how to load it on [**CrownLabs**](https://crownlabs.polito.it/) and access your [**CrownLabs**](https://crownlabs.polito.it/)' VM using SSH from a local terminal.
```

> ## How to generate an SSH key pair

Reviewer 2: Would add also here at the beginning a comment that this process needs to be done only once. Most likely, no one will ever read the last sentence.

> 1. Open a terminal and run the following command to create a [**RSA**](<https://it.wikipedia.org/wiki/RSA_(crittografia)>) key pair :

Reviewer 2: Would change the link to English documentation. Could also use `Ed25519` cryptography, but RSA is probably the default

Reviewer 4: I think that the command could be changed into `ssh-keygen -t` followed by the key type. For instance, for 2048-bit RSA:

```bash
ssh-keygen -t rsa -b 2048
```

For ED25519:

```bash
ssh-keygen -t ed25519
```

Author: We decided to omit the `-t` parameter for simplicity

   ```
   ssh-keygen -t rsa
   ```

   You can use the `-b` option to specify the length (bit size) of the key, as shown in the following example:

   ```
   ssh-keygen -b 2048 -t rsa
   ```

> 2. The command will prompt you to enter the path to the file where you want to save the key. A default path and file name similar to `/home/user_name/.ssh/id_rsa` (`C:\Users\username\.ssh\id_rsa` on **windows**) are suggested in parentheses. To accept the default path and file name (_suggested_), press **Enter**. Otherwise, enter the required path and file name, and then press **Enter**. Beware that from now on, this tutorial will assume that you have chosen the default settings.

Reviewer 2: ```suggestion
2. The command will prompt you to enter the path to the file where you want to save the key. A default path and file name similar to `/home/user_name/.ssh/id_rsa` (`C:\Users\username\.ssh\id_rsa` on **windows**) are suggested in parentheses. To accept the default path and file name (_suggested_), press **Enter**. Otherwise, enter the required path and file name, and then press **Enter**. Beware that from now on, this tutorial will assume you accepted the default settings.
```

3. A password will be asked. You are free to insert one or to ignore it.

> Now you will find in the previously chosen path (`$HOME/.ssh` if you chose the default path, `C:\Users\yourusername\.ssh`) a file with a **.pub** extension (if you chose the default file name you will find `id_rsa.pub`). This is the file containing your **SSH public key**.

Reviewer 2: ```suggestion
Now you will find in the previously chosen path (`$HOME/.ssh` or `C:\Users\yourusername\.ssh` if you kept the default one) a file with a **.pub** extension (`id_rsa.pub` as default). This is the file containing your **SSH public key**.
```

> Print the content of your **SSH public key** running:
> 
> `cat $HOME/.ssh/id_rsa.pub` or opening `C:\Users\yourusername\.ssh` with an editor in **windows**

Reviewer 2: ```suggestion
Print the content of your **SSH public key** running `cat $HOME/.ssh/id_rsa.pub` or opening `C:\Users\yourusername\.ssh` with an editor in **windows**
```

Reviewer: The equivalent in windows for `$HOME` is `%USERPROFILE%,` if you want to keep it "symmetric" 😬

Author: Thanks

> Now copy the **SSH public key**.

Reviewer 2: Would add a note that the other file contains the private key which must not be shared with anyone.

## Upload your key on [CrownLabs](https://crownlabs.polito.it/)

> Now open your page on [**CrownLabs**](https://crownlabs.polito.it/) and go in the [**Account page**](https://preprod.crownlabs.polito.it/app/account).

Reviewer 2: ```suggestion
Now open your page on [**CrownLabs**](https://crownlabs.polito.it/) and go to the [**Account page**](https://preprod.crownlabs.polito.it/app/account).
```

<br>

> 1. In the central box you will see a tab called **SSH key**. Click it to navigate in the **SSH keys** **management** panel.

Reviewer 2: ```suggestion
1. In the central box you will see a tab called **SSH key**. Click it to navigate to the **SSH keys** **management** panel.
```

![image-20211112184900823](./img/1_1.png)

<br>

> 2. Now click on **Add SSH key** to open a modal that will allow you to upload your key.

Reviewer 2: ```suggestion
2. Now click on **Add SSH key** to open a popup that will allow you to upload your key.
```

Nit. Just a little less technical.

![image-20211112185058163](./img/1_2.png)

<br>

> 3. Paste your **Public key** (previously copied) inside the modal and add a **Name** (whatever you want).

Reviewer 2: ```suggestion
3. Paste your **Public key** (previously copied) inside the popup and add a **Name** to later identify it (whatever you want).
```

![image-20211112190017165](./img/1_3.png)

<br>

> 4. Now click on **Save**, the modal will close and your **SSH public key** will be uploaded.

Reviewer 2: ```suggestion
4. Now click on **Save**, the popup will close and your **SSH public key** will be uploaded.
```

![image-20211112190234270](./img/1_4.png)

<br>

## Connect to a VM from terminal

<br>

1. Create an instance of a VM in your [**CrownLabs Dashboard page**](https://preprod.crownlabs.polito.it/app).

![image-20211112191418307](./img/2_1.png)

<br>

2. Now move to the **[Active page](https://preprod.crownlabs.polito.it/app/active)** and click on the **SSH** icon inside the row of the **VM** that you have just created. A popup with a shell command inside will be shown, **copy** that command.

![image-20211112193803509](./img/2_2.png)

<br>

3. Now open a **terminal**, paste the previously copied command and run it.
   A prompt asking if you trust the host you are trying to connect with will appear. Type **yes** and press **Enter**.

>    Then another prompt asking for a **password** will appear. Type **crownlabs** and press **Enter** to **login** inside your **VM**.

Reviewer 2: ```suggestion
   Then another prompt asking for a **password** might appear. Type **crownlabs** and press **Enter** to **login** inside your **VM**.
```

Hopefully, we will get rid of that soon.

![image-20211112194835635](./img/2_3.png)

<br>

## Conclusion

> Now you are able to log in your **VMs** using **SSH**. Note that the **generation** of the **SSH key pair** an the upload of the **SSH public key** doesn't have to be repeated each time you create a new **VM**, this process only needs to be done once.

Reviewer 2: ```suggestion
Now you are able to log in your **VMs** using **SSH**. Note that the **generation** of the **SSH key pair** and the upload of the **SSH public key** does not have to be repeated each time you create a new **VM**, this process only needs to be done once.
```

Reviewer 2: Would also add that it is possible to add other keys following the same process to access the VMs from different PCs
